File: project/.sdk/tm/lua/utility/make_request.lua
-- ProjectName SDK utility: make_request

local Response = require("core.response")
local Result = require("core.result")

local function make_request_util(ctx)
  if ctx.out["request"] ~= nil then
    return ctx.out["request"], nil
  end

  local spec = ctx.spec
  local utility = ctx.utility

  local response = Response.new({})
  local result = Result.new({})
  ctx.result = result

  if spec == nil then
    return nil, ctx:make_error("request_no_spec",
      "Expected context spec property to be defined.")
  end

  local fetchdef, err = utility.make_fetch_def(ctx)
  if err ~= nil then
    response.err = err
    ctx.response = response
    spec.step = "postrequest"
    return response, nil
  end

  if ctx.ctrl.explain ~= nil then
    ctx.ctrl.explain["fetchdef"] = fetchdef
  end

  spec.step = "prerequest"

  local url = fetchdef["url"] or ""
  local fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

  if fetch_err ~= nil then
    response.err = fetch_err
  elseif fetched == nil then
    response = Response.new({
      err = ctx:make_error("request_no_response", "response: undefined"),
    })
  elseif type(fetched) == "table" then
    response = Response.new(fetched)
  else
    response.err = ctx:make_error("request_invalid_response", "response: invalid type")
  end

  spec.step = "postrequest"
  ctx.response = response

  return response, nil
end

return make_request_util

File: project/.sdk/tm/py/utility/make_request.py
# ProjectName SDK utility: make_request

from __future__ import annotations
from core.response import ProjectNameResponse
from core.result import ProjectNameResult


def make_request_util(ctx):
    if ctx.out.get("request") is not None:
        return ctx.out["request"], None

    spec = ctx.spec
    utility = ctx.utility

    response = ProjectNameResponse({})
    result = ProjectNameResult({})
    ctx.result = result

    if spec is None:
        return None, ctx.make_error("request_no_spec",
            "Expected context spec property to be defined.")

    fetchdef, err = utility.make_fetch_def(ctx)
    if err is not None:
        response.err = err
        ctx.response = response
        spec.step = "postrequest"
        return response, None

    if ctx.ctrl.explain is not None:
        ctx.ctrl.explain["fetchdef"] = fetchdef

    spec.step = "prerequest"

    url = fetchdef.get("url", "")
    fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

    if fetch_err is not None:
        response.err = fetch_err
    elif fetched is None:
        response = ProjectNameResponse({
            "err": ctx.make_error("request_no_response", "response: undefined"),
        })
    elif isinstance(fetched, dict):
        response = ProjectNameResponse(fetched)
    else:
        response.err = ctx.make_error("request_invalid_response", "response: invalid type")

    spec.step = "postrequest"
    ctx.response = response

    return response, None
